### PR TITLE
initial implementation

### PR DESCRIPTION
**Pull request by leb4r**
## what

* initial implementation
* creates AWS budgets from `var.budgets`
* optionally creates an SNS topic and Lambda setup for Slack notifications if `var.notifications_enabled` is `true`
* optionally creates an KMS key if `var.encryption_enabled` is `true` (the default key will not work since `budgets.amazonaws.com` needs permissions to the key) 

## why
* an AWS native way to set custom budget alerts

## notes

- the only places where KMS is not used is for the CloudWatch log group. The underlying module that handles slack notifications does not expose the interface to set those. 
- if `notifications_enabled` is `true`, the module will try to create an SNS topic with a Lambda setup for Slack notifications
- if `notifications_enabled` is `true` and `encryption_enabled` is `true` then a KMS key will be created with a specific policy to allow `budgets.amazonaws.com` to use the key
- ~~I've commented out the `TestExamplesCompleteDisabled` test since it was trying to apply an empty plan instead of displaying `No changes`. After the tests ran, you can manually run terraform apply and the proper output would be show.~~

Screenshot from 2022-01-18 10-18-30

### CODE REVIEW

File: test/src/examples_complete_test.go [hunk 1]
```diff
@@ -34,45 +32,29 @@ func TestExamplesComplete(t *testing.T) {
...
+	// verify ARN of KMS key is not empty
+	kmsKeyId := terraform.Output(t, terraformOptions, "kms_key_id")
+	assert.NotEmpty(t, kmsKeyId)
 }
+
+// Test the Terraform module in examples/complete doesn't attempt to create resources with enabled=false
+// This is commented out b/c the initial plan, Terraform tries to apply, even though there is nothing to apply
+// func TestExamplesCompleteDisabled(t *testing.T) {
+// 	testNoChanges(t, "../../examples/complete")
+// }
```
Comment (korenyoni) [inline test/src/examples_complete_test.go:60]: I'm not sure, but I think you are running this plan in parallel and Terratest is running an init while the other test is running.

You can do the following to make sure the tests have separate states:

https://github.com/cloudposse/terraform-aws-ssm-tls-self-signed-cert/blob/f7bc31ea9a4f40bdf209a1c8be51bbf3932e89fb/test/src/examples_complete_test.go#L29-L51

File: README.md [hunk 2]
```diff
@@ -114,6 +114,21 @@ module "example" {
   # Cloud Posse recommends pinning every module to a specific version
   # version = "x.x.x"
 
+  budgets = [
+    {
+
+    }
+  ]
```
Comment (korenyoni) [inline README.md:121]: Did you mean to fill this in later but forgotten to get back to it?

**Review by korenyoni — COMMENTED** [review]
LGTM but notes on the README and Terratest suite

File: README.md (hunk 2 above)
Comment (leb4r) [reply README.md:121]: Sure did!

File: test/src/examples_complete_test.go (hunk 1 above)
Comment (leb4r) [reply test/src/examples_complete_test.go]: Okay, used the function from the example listed above.

**Review by korenyoni — APPROVED** [review]
Looks good to me.

**Pull request merged**